### PR TITLE
implement ipfs pin update

### PR DESCRIPTION
**Pull request by whyrusleeping**
Makes re-pinning a slightly changed dataset much faster

License: MIT
Signed-off-by: Jeromy <jeromyj@gmail.com>

### CODE REVIEW

**Comment by whyrusleeping** [comment]
I'll want a solid review of the algorithm here

File: merkledag/utils/diffenum.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	if err != nil {
+		return fmt.Errorf("get %s: %s", to, err)
+	}
+
+	diff := getLinkDiff(fnd, tnd)
+
+	sset := cid.NewSet()
+	for _, c := range diff {
+		if c.a == nil {
+			err := mdag.EnumerateChildrenAsync(ctx, mdag.GetLinksDirect(dserv), c.b, sset.Visit)
```
Comment (Kubuxu) [inline merkledag/utils/diffenum.go]: This line is really long

File: merkledag/utils/diffenum.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+type diffpair struct {
+	a, b *cid.Cid
```
Comment (Kubuxu) [inline merkledag/utils/diffenum.go]: I would rename it to `old, new`. It clears up a lot.

File: merkledag/utils/diffenum.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+func getLinkDiff(a, b node.Node) []diffpair {
```
Comment (Kubuxu) [inline merkledag/utils/diffenum.go]: It took me a minute to properly get what this function does but now I understand.

It diffs links as a set, it doesn't matter for it if some link is repeated.

File: merkledag/utils/diffenum.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+	sset := cid.NewSet()
+	for _, c := range diff {
+		if c.a == nil {
+			err := mdag.EnumerateChildrenAsync(ctx, mdag.GetLinksDirect(dserv), c.b, sset.Visit)
+			if err != nil {
+				return err
+			}
+		} else {
+			err := DiffEnumerate(ctx, dserv, c.a, c.b)
```
Comment (Kubuxu) [inline merkledag/utils/diffenum.go]: Are we worried here about too deep recursion? Should it be turned into loop with TODO list instead?

Comment (Kubuxu) [inline merkledag/utils/diffenum.go]: Also coverage shows that this isn't tested at all.

File: core/commands/pin.go [hunk 5]
```diff
@@ -332,6 +333,83 @@ Example:
...
+new pin and removing the old one.
+`,
+	},
+
+	Arguments: []cmds.Argument{
+		cmds.StringArg("from-path", true, false, "Path to old object."),
+		cmds.StringArg("to-path", true, false, "Path to new object to be pinned."),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption("unpin", "Remove the old pin.").Default(true),
```
Comment (whyrusleeping) [inline core/commands/pin.go]: Its if you want to use the optimized pinning functionality and still keep the old pin

File: merkledag/utils/diffenum.go (hunk 1 above)
Comment (whyrusleeping) [inline merkledag/utils/diffenum.go]: It is really tho?

File: core/commands/pin.go (hunk 5 above)
Comment (Kubuxu) [inline core/commands/pin.go]: Ahh, right, because it will piggy bank on the already pinned hashes.

**Comment by kevina** [comment]
@whyrusleeping I would like to review this but it may take up to week before I can get to it.

File: merkledag/utils/diffenum.go (hunk 2 above)
Comment (whyrusleeping) [inline merkledag/utils/diffenum.go]: cant use 'new' lol, its a keyword

Comment (Kubuxu) [inline merkledag/utils/diffenum.go]: ouch, annoying

File: merkledag/utils/diffenum.go (hunk 1 above)
Comment (Kubuxu) [inline merkledag/utils/diffenum.go]: aka. too wide for my vim setup.

File: merkledag/utils/diffenum.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	}
+
+	return nil
+}
+
+type diffpair struct {
+	bef, aft *cid.Cid
+}
+
+// getLinkDiff returns a changset (minimum edit distance style) between nodes
```
Comment (kevina) [inline merkledag/utils/diffenum.go]: I was rather confused by what this is doing.  From the code it is fairly clear but the phrase "minimum edit distance style" confused me.  As someone who wrote a spell checker I was expecting a traditional edit dist algo, and was very confused my the use of the maps.  I would probably have had an easier time understanding what is going on if I wasn't trying to reconcile the code with what a traditional string edit dist. algo would be doing.

File: pin/pin.go [hunk 7]
```diff
@@ -86,6 +87,11 @@ type Pinner interface {
 	Pin(context.Context, node.Node, bool) error
 	Unpin(context.Context, *cid.Cid, bool) error
 
+	// Update updates a recursive pin from one cid to another
+	// this is more efficient than simply pinning the new one and unpinning the
+	// old one
+	Update(context.Context, *cid.Cid, *cid.Cid, bool) error
```
Comment (kevina) [inline pin/pin.go]: I would name the parameters here.  The use of the bool in Update differs from the use in pin and unpin.

**Pull request merged**